sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,455 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/usb_logging.h"
+
+#include <ctype.h>
+
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/pinmux_testutils.h"
+#include "sw/device/lib/testing/usb_testutils.h"
+#include "sw/device/lib/testing/usb_testutils_controlep.h"
+
+#define MODULE_ID MAKE_MODULE_ID('u', 's', 'l')
+
+// Size of each logging buffer; data is collected in a buffer until the buffer
+// is full (or a timeout occurs) and is then transmitted as a single transfer
+// at higher throughput.
+#define USBLOG_BUF_SIZE 0x1000u
+
+// Simple double-buffering scheme allows us to achieve higher throughput via
+// the existing usb_testutils layer and usbdev.
+#define USBLOG_NUM_BUFFERS 2u
+
+// Maximum length of the configuration descriptor.
+#define CFG_DSCR_LEN_MAX \
+  (USB_CFG_DSCR_LEN +    \
+   USBLOG_MAX_STREAMS * (USB_INTERFACE_DSCR_LEN + 1u * USB_EP_DSCR_LEN))
```

Comment:
Use `enum` constants, per the style guide?

Reply:
Okay, I was mimicking the existing usbdev-related code. I'll try to ensure that code outside of usb_logging is revised over time.